### PR TITLE
Docs: Add comprehensive troubleshooting guide

### PR DESCRIPTION
**Pull request by khlaifiabilel**
## Summary
This PR adds a comprehensive troubleshooting guide to help users resolve common issues encountered when using TorchTitan.

## Motivation
- Users frequently encounter similar issues (OOM errors, setup problems, configuration errors)
- Reduces repetitive questions on PyTorch forums and GitHub issues  
- Provides a central reference for debugging and problem-solving
- Makes it easier for new users to get started

### Setup & Installation
- GPU detection issues and PyTorch CUDA configuration
- Tokenizer download failures and HuggingFace authentication
- Import errors and PYTHONPATH configuration
- Pre-commit hook installation and troubleshooting

### Training Issues
- Out of Memory (OOM) errors with multiple solutions (batch size, activation checkpointing, gradient accumulation, CPU offload)
- Training hangs and debugging with environment variables
- Checkpoint loading failures and compatibility issues
- Loss divergence (NaN/Inf) with solutions for learning rate, gradient clipping, and deterministic debugging

### Configuration Issues
- Boolean flag overrides from CLI (using --no prefix)
- Config validation errors and debugging
- Parallelism dimension conflicts with clear examples

### Distributed Training
- Multi-node setup issues (network configuration, master address/port)
- NCCL communication errors and timeout settings
- Network interface configuration

### Performance Issues
- Low GPU utilization diagnosis and solutions
- Slow training speed optimization tips
- Profiling commands and benchmarking guidance

### Getting Help
- Where to ask questions (Forums, GitHub Issues, Discussions)
- How to report issues effectively with required information
- Useful debugging commands for quick diagnosis
- Links to related documentation

## Testing
- [x] All commands have been verified to work
- [x] Pre-commit hooks pass (trailing whitespace fixed)
- [x] Cross-references to other docs are correct (debugging.md, fsdp.md, checkpoint.md)
- [x] Markdown formatting is valid

## Type of Change
- [x] Documentation update

## Additional Context
This guide complements existing documentation by providing quick, actionable solutions to common problems. It's designed to be easily searchable and scannable for users who are stuck. The guide focuses on practical solutions with copy-pasteable commands and clear examples.

Future enhancements could include:
- Hardware-specific troubleshooting (AMD GPUs, different CUDA versions)
- Visual flowcharts for issue diagnosis
- Cloud provider-specific setup guides
- Video tutorials for complex multi-node setups

## Checklist
- [x] Documentation follows the style of this project
- [x] Pre-commit hooks pass
- [x] All links are valid and working
- [x] Cross-references to other docs are correct
- [x] I have read the CONTRIBUTING.md document

### CODE REVIEW

File: docs/troubleshooting.md [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+   [training]
+   batch_size = 1
+   global_batch_size = 8  # Effective batch size through accumulation
+   ```
+
+4. **Enable FSDP2 with CPU offload**:
+   ```toml
+   [model]
+   fsdp_cpu_offload = true
+   ```
```
Comment (fegin) [inline docs/troubleshooting.md:142]: There are other approaches to do first like TP and effectively reducing the global batch size. CPU offload can be very slow and should be the last resort?

File: docs/troubleshooting.md [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+3. **Verify network connectivity** (multi-node):
+   ```bash
+   # Test from node 0 to node 1
+   ping <node1-ip>
+   ```
+
+4. **Start with debug model**:
+   ```bash
+   CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" ./run_train.sh
+   ```
```
Comment (fegin) [inline docs/troubleshooting.md:178]: Maybe add `py-spy` to understand which rank is doing what.

File: docs/troubleshooting.md [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+   ```bash
+   CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh \
+     --profiling.enable_memory_snapshot \
+     --profiling.save_memory_snapshot_folder memory_snapshot
+   ```
+   Then visualize at https://pytorch.org/memory_viz
+
+### Training Hangs at Start
+
+**Problem:** Training process hangs without error messages
```
Comment (fegin) [inline docs/troubleshooting.md:154]: Sometimes, it is because we only output rank0 log and cause some ranks errors are swallowed by TorchRun. `LOG_RANK=0,1,2,3,4,5,6,7` can help. I used this a lot.

File: docs/troubleshooting.md [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+
+1. **Verify checkpoint path** in config:
+   ```toml
+   [checkpoint]
+   folder = "./outputs/checkpoint"
+   ```
+
+2. **Check checkpoint compatibility**:
+   - Ensure same model architecture
+   - Verify parallelism settings match
```
Comment (fegin) [inline docs/troubleshooting.md:197]: Parallelism settings match is not required.

File: docs/troubleshooting.md [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+
+### NCCL Errors
+
+**Problem:** NCCL communication errors
+
+**Solutions:**
+
+1. **Set NCCL timeout**:
+   ```bash
+   export NCCL_TIMEOUT=1800
```
Comment (fegin) [inline docs/troubleshooting.md:339]: TorchTitan has a comm timeout setting. We should just use that?

File: docs/troubleshooting.md [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+   batch_size = 4  # Increase if you have memory headroom
+   ```
+
+2. **Enable torch.compile**:
+   ```toml
+   [model]
+   compile = true
+   ```
+
+3. **Check data loading** isn't the bottleneck:
```
Comment (fegin) [inline docs/troubleshooting.md:374]: Have you ever encountered this case when using TorchTitan or have you heard such a case?

File: docs/troubleshooting.md [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+   ```toml
+   [model]
+   compile = true
+
+   [activation_checkpoint]
+   mode = 'selective'  # Balance memory and speed
+   ```
+
+2. **Check TFLOPs and MFU** in logs:
+   - Compare with benchmark numbers in `benchmarks/`
```
Comment (fegin) [inline docs/troubleshooting.md:400]: And check the trace to understand if there are exposed communications.

File: docs/troubleshooting.md [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+nvidia-smi --list-gpus
+
+# Monitor GPU usage during training
+watch -n 1 nvidia-smi
+
+# Check config interpretation
+python -m torchtitan.config.manager --job.config_file config.toml
+
+# Run with maximum debugging
+export TORCH_DISTRIBUTED_DEBUG=DETAIL
```
Comment (fegin) [inline docs/troubleshooting.md:455]: I'm not sure if we want to advocate this. Given that we now have NCCL flight recorder. It is usually easier to just check flight recorder result.

Comment (fegin) [reply docs/troubleshooting.md:455]: One most common use case people turn on these debug information is NCCL timeout. We should just focus on `NCCL flight recorder`. Other debug log is likely too expert-only information.

File: docs/troubleshooting.md (hunk 1 above)
Comment (wwwjn) [reply docs/troubleshooting.md:142]: Agree, to resolve OOM issue, it's a very case-by-case decision and a general guide here might not be that helpful

File: docs/troubleshooting.md [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+   export TORCH_DISTRIBUTED_DEBUG=DETAIL
+   export NCCL_DEBUG=INFO
+   ```
+
+2. **Check GPU availability** on all nodes:
+   ```bash
+   nvidia-smi
+   ```
+
+3. **Verify network connectivity** (multi-node):
```
Comment (wwwjn) [inline docs/troubleshooting.md:169]: This part is not even related to torchtitan

File: docs/troubleshooting.md [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+
+3. **Use checkpoint conversion** if from HuggingFace:
+   ```bash
+   python scripts/checkpoint_conversion/convert_from_hf.py \
+     --model_name llama3 \
+     --hf_checkpoint_path /path/to/hf \
+     --output_path ./checkpoint
+   ```
+
+### Loss is NaN or Diverging
```
Comment (wwwjn) [inline docs/troubleshooting.md:207]: Also this issue might be very different under different cases: Simpler case might be learning rate is too high,  but sometimes it even because of a pytorch or torchtitan bug

File: docs/troubleshooting.md [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+   export MASTER_PORT=29500
+   ```
+
+2. **Verify all nodes can communicate**:
+   ```bash
+   # From each node, ping master
+   ping $MASTER_ADDR
+   ```
+
+3. **Check firewall settings** - ensure port is open
```
Comment (wwwjn) [inline docs/troubleshooting.md:319]: This part does not seem to be related to torchtitan

File: docs/troubleshooting.md [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+# WRONG - This won't work
+--profiling.enable_memory_snapshot=false
+
+# CORRECT - Use --no prefix
+--profiling.no_enable_memory_snapshot
+# or
+--profiling.no-enable-memory-snapshot
+```
+
+See [debugging.md](debugging.md#overriding-boolean-flags-from-toml-via-cli) for more details.
```
Comment (wwwjn) [inline docs/troubleshooting.md:256]: As you mentioned, this part is overlapped with debugging.md

**Review by wwwjn — COMMENTED** [review]
Thanks for the contribution, I think it's a good guide for torchtitan new users! However, I think it's better to be made into a blog / tutorial, but not put into torchtitan/docs part, because of following reasons:
1. In torchtitan docs section, the instructions are divided into different topics, and this guide has a lot overlap with other docs.
2. There are basic GPU usage knowledge, model training knowledge, which is not specific related to torchtian. 
3. Sometimes debugging is highly case-by-case, providing general starting point is good but might not be helpful in the end.

I'm open to discuss where we should put our docs and how to organize them.

File: docs/troubleshooting.md (hunk 10 above)
Comment (tianyu-l) [reply docs/troubleshooting.md:207]: Maybe also good to add how to enable anomaly detection.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
This is a nice doc to have.

But I have the worry that this document explicitly mentions a lot of configs, which would be easy to break, or significantly slow down the development / refactor.

I don't want the torchtitan core team to maintain this doc. Maybe it should be part of the broader releasing engineering's job.

cc @dcci @svekars

**Comment by dcci** [comment]
> This is a nice doc to have.
> 
> But I have the worry that this document explicitly mentions a lot of configs, which would be easy to break, or significantly slow down the development / refactor.
> 
> I don't want the torchtitan core team to maintain this doc. Maybe it should be part of the broader releasing engineering's job.
> 
> cc @dcci @svekars

Agree. The maintenance cost of this is relatively high.

**Comment by khlaifiabilel** [comment]
Thank you for the feedback. I understand the concerns about overlap and the long-term maintenance cost of this guide. I am closing this PR rather than adding that burden to the core documentation. I may preserve the material as an external tutorial and will discuss the scope with maintainers before proposing future documentation changes.
